lib/buffer.js

```diff
@@ -24,82 +24,148 @@ function createPool() {
 }
 createPool();
 
+function Buffer(arg) {
+  if (!(this instanceof Buffer)) {
+    // Avoid going through an ArgumentsAdaptorTrampoline in the common case.
+    if (arguments.length > 1)
+      return new Buffer(arg, arguments[1]);
 
-function Buffer(subject, encoding) {
-  if (!(this instanceof Buffer))
-    return new Buffer(subject, encoding);
-
-  if (typeof subject === 'number') {
-    this.length = +subject;
-
-  } else if (typeof subject === 'string') {
-    if (typeof encoding !== 'string' || encoding.length === 0)
-      encoding = 'utf8';
-    this.length = Buffer.byteLength(subject, encoding);
+    return new Buffer(arg);
+  }
 
-  // Handle Arrays, Buffers, Uint8Arrays or JSON.
-  } else if (subject !== null && typeof subject === 'object') {
-    if (subject.type === 'Buffer' && Array.isArray(subject.data))
-      subject = subject.data;
-    this.length = +subject.length;
+  this.length = 0;
+  this.parent = undefined;
 
-  } else {
-    throw new TypeError('must start with number, buffer, array or string');
+  // Common case.
+  if (typeof(arg) === 'number') {
+    fromNumber(this, arg);
+    return;
   }
 
-  if (this.length > kMaxLength) {
-    throw new RangeError('Attempt to allocate Buffer larger than maximum ' +
-                         'size: 0x' + kMaxLength.toString(16) + ' bytes');
+  // Slightly less common case.
+  if (typeof(arg) === 'string') {
+    fromString(this, arg, arguments.length > 1 ? arguments[1] : 'utf8');
```

Comment:
Two reasons, one aesthetic, one technical.  Aesthetic: it matches the check at the top of the function.  Technical: an unguarded `arguments[1]` is megamorphic whereas the length check keeps the property/index lookups monomorphic.

I didn't benchmark it to the death but I did look at the generated code and came to the conclusion that monomorphic is best, even with the additional property lookup.  I don't know if you have looked at the machine code for KeyedLoadIC_Megamorphic but it's pretty complex.